File: Data Structures & Algorithms/course-schedule/submission-4.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        pre = defaultdict(list)

        for u, v in prerequisites:
            pre[v].append(u)
        
        visited = set()

        def dfs(crs):
            if crs in visited:
                return False
            
            if pre[crs] == []:
                return True
            
            visited.add(crs)
            for v in pre[crs]:
                if not dfs(v):
                    return False
            
            visited.remove(crs)
            pre[crs] = []
            return True

        for r in range(numCourses):
            if not dfs(r):
                return False
        
        return True

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] pre = new int[n];
        pre[0] = 1;

        for (int i = 1; i < nums.length; i++) {
            pre[i] = pre[i-1] * nums[i-1];
        }

        int post = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            pre[i] = post * pre[i];
            post *= nums[i];
        }

        return pre;
    }
}  
